### DISCUSSION TITLE
EIP 4337 - Snaps

### DISCUSSION THREAD

**Discussion by Riki0923** [MetaMask/snaps#1379] [Ideas]
Hello Everyone.

I am having a team where we would like to implement erc 4337 to Metamask Snaps. Our only question would be that is it possible to convert user's transactions into a userOp ( UserOperation  ) or not? Because basically this is the only thing which decides whether we can do our project or not?

Thank you.

Regards,

Krisztian

**Comment by Montoya** [comment]
If you are managing an Ethereum account on behalf of a user, you can handle transactions for those users however you want. You can implement all of the wallet logic in your snap to submit userOps. 

However, you cannot implement custom transaction handling for the Ethereum accounts managed by MetaMask. There is no API to do this yet.

**Reply by cwhcheng** [reply]
i follow this repo https://github.com/MetaMask/snap-account-abstraction-keyring/tree/v0.2.2 to create a AA account in snap and I can connect the AA ac to any dapp. But when i actually try to send out a transaction (let's say doing a swap in Uniswap), it throws me an error `Invalid options to add user operation smartContractAccount - Expected an object, but received: undefined { "originalError": {} }`  Do you know how to deal with this issue? thanks!

**Comment by Riki0923** [comment]
Hey.

It is probably the UserOperation, which is an object that needs to be
passed in order to make an accout abstracted transaction.

Regards,

Krisztian

**Reply by cwhcheng** [reply]
do you know how to pass userOp object? cuz in the `packages/snap/src/index.ts`, i try to print out log on `onRpcRequest`, `onKeyringRequest`, `onTransaction`, but i cannot get any log msg, so i suspect it never go into the snap and i don't know why. Appreciate your help!

**Reply by Montoya** [reply]
Here is a guide on how to view the console log messages: https://docs.metamask.io/snaps/how-to/debug-a-snap/

**Reply by cwhcheng** [reply]
thanks! I am able to get log msg for using normal account but I cannot get any custom log for using AA account. I only receive error msg `Invalid options to add user operation smartContractAccount - Expected an object, but received: undefined { "originalError": {} }` and `eth_sendTransaction -> error` in `background.html`

**Reply by cwhcheng** [reply]
I wanna know is it possible to directly use an AA account in the snap to sign and confirm a UserOperation inside Metamask (without jumping to another website) when interacting normal dapp like uniswap or aave?
